File: olympic-hcm/2015-lop10/positive/positive.cpp
#include <iostream>
#include <vector>
#include <queue>

#define input_file "positive.inp"
#define output_file "positive.out"

using namespace std;

struct cell
{
    int r;
    int c;
};

int R, C;
vector<vector<int>> grid;
vector<vector<bool>> visited;

// 8 hướng di chuyển
vector<cell> directions = {
    {-1, 0}, // lên
    {0, 1}, // phải
    {1, 0}, // xuống
    {0, -1}, // trái
    {-1, 1}, // phải trên     
    {1, 1}, // phải dưới     
    {1, -1}, // trái dưới    
    {-1, -1} // trái trên
    };

int region_count = 0;


void input()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    freopen(input_file, "r", stdin);

    cin >> R >> C;

    grid.resize(R, vector<int>(C));

    for (int r = 0; r < R; ++r)
    {
        for (int c = 0; c < C; ++c)
        {
            cin >> grid[r][c];
        }
    }
}


void bfs(cell start)
{
    // Khởi tạo hàng đợi q
    queue<cell> q;
    q.push(start);

    cell current; // ô đang xét
    cell next;    // ô lân cận

    // Duyệt lưới grid theo BFS và đánh dấu đã duyệt cho tất cả ô trong cùng một vùng
    while (!q.empty())
    {
        current.r = q.front().r;
        current.c = q.front().c;
        q.pop();

        visited[current.r][current.c] = true;

        for (int d = 0; d < directions.size(); ++d)
        {
            // Lấy giá trị của ô next
            next.r = current.r + directions[d].r;
            next.c = current.c + directions[d].c;

            // Kiểm tra xem ô next có thoả các điều kiện sau không:
            // 1. nằm trong lưới
            // 2. có giá trị dương
            // 3. chưa duyệt
            if (next.r > -1)
                if (next.r < R)
                    if (next.c > -1)
                        if (next.c < C)
                            if (grid[next.r][next.c] > 0)
                                if (visited[next.r][next.c] == false)
                                {
                                    // Đánh dấu ô next đã duyệt
                                    visited[next.r][current.c] = true;

                                    // Đẩy ô next vào hàng đợi
                                    q.push(next);
                                }
        }
    }
}


void process()
{
    // Khởi tạo mảng visited
    visited.resize(R, vector<bool>(C, false));

    cell start_cell;

    // Duyệt từng ô trong lưới grid
    for (int r = 0; r < R; ++r)
    {
        for (int c = 0; c < C; ++c)
        {
            // Nếu ô có giá trị dương và chưa duyệt
            if (grid[r][c] > 0)
            {
                if (visited[r][c] == false)
                {
                    // thì bắt đầu tính là một vùng mới
                    region_count += 1;

                    // thực hiện bfs từ ô này
                    start_cell.r = r;
                    start_cell.c = c;
                    bfs(start_cell);
                }
            }
        }
    }
}


void output()
{
    freopen(output_file, "w", stdout);
    cout << region_count;
}


int main()
{
    input();
    process();
    output();

    return 0;
}

File: olympic-hcm/2015-olympic10/positive/positive.py
import os
import sys
from collections import deque

input_file = os.path.join(sys.path[0], 'positive.inp')
output_file = os.path.join(sys.path[0], 'positive.out')

class cell:
    def __init__(self, r, c):
        self.r = r
        self.c = c

R, C = 0, 0
grid = []
visited = []

# 8 hướng di chuyển
directions = [
    cell(-1, 0), # lên
    cell(0, 1), # phải
    cell(1, 0), # xuống
    cell(0, -1), # trái
    cell(-1, 1), # phải trên
    cell(1, 1), # phải dưới 
    cell(1, -1), # trái dưới   
    cell(-1, -1) # trái trên
    ]

region_count = 0


def input_data():
    global R, C, grid
    
    with open(input_file) as f:
        R, C = map(int, f.readline().split())
        grid = [list(map(int, f.readline().split())) for _ in range(R)]


def bfs(start):
    global R, C, grid, visited

    # Khởi tạo hàng đợi q
    q = deque([start])

    # Duyệt lưới grid theo BFS và đánh dấu đã duyệt cho tất cả ô trong cùng một vùng
    while q:
        current = q.popleft()
        visited[current.r][current.c] = True

        for direction in directions:
            # Lấy giá trị của ô next
            next_r = current.r + direction.r
            next_c = current.c + direction.c

            # Kiểm tra xem ô next có thoả các điều kiện sau không:
            # 1. nằm trong lưới
            # 2. có giá trị dương
            # 3. chưa duyệt
            if next_r > -1:
                if next_r < R:
                    if next_c > -1:
                        if next_c < C:
                            if grid[next_r][next_c] > 0:
                                if visited[next_r][next_c] == False:
                                    # Đánh dấu ô next đã duyệt
                                    visited[next_r][next_c] = True

                                    # Đẩy ô next vào hàng đợi
                                    q.append(cell(next_r, next_c))


def process():
    global grid, region_count, visited
    
    # Khởi tạo mảng visited
    visited = [[False] * C for _ in range(R)]

    # Duyệt từng ô trong lưới grid
    for r in range(R):
        for c in range(C):
            # Nếu ô có giá trị dương và chưa duyệt
            if grid[r][c] > 0:
                if visited[r][c] == False:
                    # thì bắt đầu tính là một vùng mới
                    region_count += 1

                    # thực hiện bfs từ ô này
                    bfs(cell(r, c))


def output():
    global region_count

    with open(output_file, "w") as f:
        f.write(f'{region_count}')


if __name__ == '__main__':
    input_data()
    process()
    output()

